File: 原始代码/延迟求和算法的Demo/Beamforming_DAS_Demo.m
% ----------- DAS 波束成像算法（扫频模式）
% ------ 可以设置多个不同频率、不同声压级的点声源
% ------ 可以设置不同距离的扫描平面
% ------ 可以选择不同的麦克风阵列
% ------ 可以设置想要搜索的频段
% ------ 可以调整网格分辨率

close all; clear; clc;

% 添加路径
addpath('.\Algorithm')
addpath('.\Prepocess')

% 麦克风阵列限定区域
mic_x = [-0.5 0.5];
mic_y = [-0.5 0.5];

% 扫描声源限定区域
scan_x = [-3 3];
scan_y = [-3 3];

% 麦克风阵列平面与扫描屏幕的距离
z_source = 3;

% 声速
c = 343; 

% 信号的采样频率
fs = 8000;

% 导入麦克风阵列
load 56_spiral_array.mat
mic_x_axis = array(:,1); mic_y_axis = array(:,2); mic_z_axis = 0;
mic_pos = [mic_x_axis mic_y_axis]; mic_pos(:,3) = mic_z_axis;
mic_centre = mean(mic_pos); % 阵列中心的坐标

% 绘制麦克风阵列
figure;
plot(mic_x_axis, mic_y_axis,'k.', 'MarkerSize',20);
xlim([min(mic_x_axis)-0.1, max(mic_x_axis)+0.1])
ylim([min(mic_y_axis)-0.1, max(mic_y_axis)+0.1])

% 构建声源点
source_x = [-1,0.5]';
source_y = [0,1]';

% 设定声源频率
source1_freq = 2000;  source2_freq = 3000;
sources_freq = [source1_freq, source2_freq]';

% 设定信号持续时间
t_start = 0;  t_end = 0.02;
source_duration = t_end*ones(length(source_x),1);

% 设定声源声压有效值
source1_spl = 100; source2_spl = 100;  %注意两个不要差太远
sources_spl = [source1_spl, source2_spl].';

% 整合声源信息：source_info
% 声源点坐标x / 声源点坐标y / 声源点坐标z（到扫描平面距离）/ 声源频率/ 声压值
source_info = [source_x, source_y, z_source*ones(length(source_x),1), sources_freq, sources_spl, source_duration]; 

% 获取麦克风阵列输出
mic_signal = simulateArraydata(source_info, mic_pos, c, fs, source_duration, mic_centre);  %

% 确定扫描频段（800-4000 Hz）
search_freql = 800;
search_frequ = 4000;

% 计算CSM以及确定扫描频率
[CSM, freqs] = developCSM(mic_signal.', search_freql, search_frequ, fs, t_start,t_end);

% 扫描网格的分辨率
scan_resolution = 0.1;

% 计算转向矢量
g = steerVector(z_source, freqs, [scan_x scan_y], scan_resolution, mic_pos.', c, mic_centre);

% 波束成像 -- DAS算法
[X, Y, B] = DAS(CSM, g, freqs, [scan_x scan_y], scan_resolution);

% 声压级单位转换
B(B<0)=0;
SPL = 20*log10((eps+sqrt(real(B)))/2e-5);

% 绘制波束成像图
figure;
BF_dr = 6; maxSPL = ceil(max(SPL(:)));
contourf(X, Y, SPL, (maxSPL-BF_dr):1:maxSPL); colorbar;
hold on; plot(source_x(:),source_y(:),'r*'); 

File: 修改代码/Beamforming_DAS_demo.py
import sys
import h5py
import matplotlib.pyplot as plt
import numpy as np
import numpy.matlib
from preprocess.developCSM import developCSM, freqs_precaulate
from preprocess.steerVector import steerVector
from Algorithm.DAS import DAS
from preprocess.simulateArraydata import simulateArraydata
import wave
import scipy.io as scio
import time


def get_micSignal_from_wav(wav_path):
    '''从wav文件中获得micsignal
    自动进行裁切(8通道到6通道)
    mic_sigal.size=channel,frames_total'''
    wavfile = wave.open(wav_path)
    framerate = wavfile.getframerate()
    params = wavfile.getparams()
    nchannels, sampwidth, framerate, nframes = params[:4]
    strData = wavfile.readframes(nframes)  # 读取音频，字符串格式
    waveData = np.fromstring(strData, dtype=np.int16)  # 将字符串转化为int
    # waveData = waveData*1.0/(max(abs(waveData)))  # wave幅值归一化
    waveData = np.reshape(waveData, [nchannels, nframes])
    wavfile.close()

    mic_signal = TrackAlignment(waveData)
    mic_signal = mic_signal.T  # mic_sigal.size=channel,frames_total
    return framerate, nframes, mic_signal


def TrackAlignment(data):
    '''裁切'''
    y = data.T
    s_x = np.argsort(np.max(y, axis=0))
    y2 = []
    if abs(s_x[0]-s_x[1]) == 1:
        s_x_max = max(s_x[0:1])
        if(s_x_max == 1):
            y2 = y[:, 2:8]
        elif(s_x_max < 7 and s_x_max > 1):
            y2 = np.hstack((y[:, s_x_max+1:8], y[:, 0:s_x_max-1]))
        else:
            y2 = y[:, 0:6]
    else:
        y2 = y[:, 1:7]
    return y2


def simulateMicsignal(mic_pos, z_source, c, framerate, mic_centre, t_start, t_end):
    '''构建虚拟声源点'''
    # % 构建声源点  %注:设定信号持续时间和整合声源信息：source_info
    source_x = np.array([-0.2, 0.3]).reshape(1, -1).T  # source_x = [-1,0.5]';
    source_y = np.array([0, 0.1]).reshape(1, -1).T  # source_y = [0,1]';

    # % 设定声源频率
    source1_freq = 2000
    source2_freq = 3000  # source1_freq = 2000;  source2_freq = 3000
    # sources_freq = [source1_freq, source2_freq]';  %注:整合声源信息：source_info
    sources_freq = np.array([source1_freq, source2_freq]).reshape(1, -1).T

    # % 设定信号持续时间
    # source_duration = t_end*ones(length(source_x), 1)
    # %注: 整合声源信息：source_info和获取麦克风阵列输出simulateArraydata
    source_duration = t_end*np.ones((max(source_x.shape), 1))

    # % 设定声源声压有效值
    source1_spl = 100
    source2_spl = 100  # source1_spl = 100; source2_spl = 100
    # sources_spl = [source1_spl, source2_spl].';   %注:整合声源信息：source_info
    sources_spl = np.array([source1_spl, source2_spl]).reshape(1, -1).T

    # % 整合声源信息：source_info % 注: 获取麦克风阵列输出simulateArraydata
    # % 声源点坐标x / 声源点坐标y / 声源点坐标z（到扫描平面距离）/ 声源频率 / 声压值

    # source_info = [source_x, source_y, z_source *
    #                ones(length(source_x), 1), sources_freq, sources_spl, source_duration]
    source_info = np.concatenate((source_x, source_y, z_source*np.ones(
        (max(source_x.shape), 1)), sources_freq, sources_spl, source_duration), axis=1)
    # https://www.cnblogs.com/cymwill/p/8358866.html

    # % 获取麦克风阵列输出 % 注:计算CSM以及确定扫描频率developCSM
    mic_signal = simulateArraydata(
        source_info, mic_pos, c, framerate, source_duration, mic_centre)
    # mic_signal.size() = (nchannels, nframes) 且不归一化
    return mic_signal


def get_MicSignal():
    return


def save_wav():
    return


def get_micArray(path_full):
    try:
        darray = scio.loadmat(path_full)
    except:
        darray = h5py.File(path_full)  # 如果python 报错
    array = darray['array'][:]
    mic_x_axis = array[:, 0]
    mic_y_axis = array[:, 1]
    mic_z_axis = 0
    mic_pos = np.transpose([mic_x_axis, mic_y_axis])
    mic_pos = np.concatenate(
        (mic_pos, np.ones((mic_x_axis.size, 1))*mic_z_axis), axis=1)
    # mic_centre = mean(mic_pos); % 阵列中心的坐标
    mic_centre = mic_pos.mean(axis=0).reshape(1, -1)
    return mic_pos, mic_centre, mic_x_axis, mic_y_axis


def draw_mic_array(mic_x_axis, mic_y_axis):
    '''绘制麦克风阵列'''
    plt.figure()
    plt.plot(mic_x_axis, mic_y_axis, 'k.', markersize=20)
    plt.xlim([min(mic_x_axis)-0.1, max(mic_x_axis)+0.1])
    plt.ylim([min(mic_y_axis)-0.1, max(mic_y_axis)+0.1])
    plt.show()


def plot_figure(X, Y, SPL):
    # % 绘制波束成像图

    # 解决中文显示问题
    plt.rcParams['font.sans-serif'] = ['SimHei']
    plt.rcParams['axes.unicode_minus'] = False

    plt.figure()  # figure;
    BF_dr = 6
    # BF_dr = 6; maxSPL = ceil(max(SPL(:)))
    maxSPL = np.ceil(max(SPL.flatten('F')))
    X, Y = np.meshgrid(X, Y)
    plt.contourf(X, Y, SPL, np.arange((maxSPL-BF_dr), maxSPL+1, 1))
    plt.colorbar()
    plt.xlabel('x轴(m)')
    plt.ylabel('y轴(m)')
    plt.title('波束成像图')
    plt.show()
